certora/specs/TimelockController.spec

```
@@ -128,151 +166,91 @@ rule minDelayOnlyChange(env e) {
-    uint8 stateAfter = state(id);
-
-    assert stateBefore == PENDING() && stateAfter == DONE(), "State transition violation";
-    assert isOperationReadyBefore, "Execute before ready";
-    assert otherTimestamp != getTimestamp(otherId) => id == otherId, "Other proposal affected";
-}
-
-/*
-┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
-│ Rule: cancel correctly updates the state                                                                            │
-└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
-*/
-rule cancel(env e, bytes32 id) {
-    bytes32 otherId;
-    uint256 otherTimestamp = getTimestamp(otherId);
-
-    uint8 stateBefore = state(id);
-    cancel(e, id);
-    uint8 stateAfter = state(id);
-
-    assert stateBefore == PENDING() && stateAfter == UNSET(), "State transition violation";
-    assert otherTimestamp != getTimestamp(otherId) => id == otherId, "Other proposal affected";
-}
-
-/*
-┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
-│ Access control: only users with the PROPOSER_ROLE can call schedule or scheduleBatch                                │
+│ Rule: schedule liveliness and effets                                                                                │
 └─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
 */
-rule onlyProposer(env e, bytes32 id, method f) filtered { f ->
+rule schedule(env e, method f, bytes32 id, uint256 delay) filtered { f ->
     f.selector == schedule(address, uint256, bytes, bytes32, bytes32, uint256).selector ||
     f.selector == scheduleBatch(address[], uint256[], bytes[], bytes32, bytes32, uint256).selector
 } {
-    bool isProposer = hasRole(PROPOSER_ROLE(), e.msg.sender);
-
-    calldataarg args;
-    f@withrevert(e, args);
-
-    assert !isProposer => lastReverted, "Only proposer can propose";
+    require nonpayable(e);
+    require e.block.timestamp > 1; // Sanity
+    require e.block.timestamp + delay < max_uint256; // Sanity
+    require e.block.timestamp + getMinDelay() < max_uint256; // Sanity
+
+    bytes32 otherId; uint256 otherTimestamp = getTimestamp(otherId);
+
+    uint8 stateBefore       = state(id);
+    bool  isDelaySufficient = delay >= getMinDelay();
+    bool  isProposerBefore  = hasRole(PROPOSER_ROLE(), e.msg.sender);
+
+    helperScheduleWithRevert(e, f, id, delay);
+
+    if (lastReverted) {
+        assert stateBefore != UNSET() || !isDelaySufficient || !isProposerBefore, "Unexpected revert";
+    } else {
+        assert stateBefore == UNSET() && state(id) == PENDING(), "State transition violation";
+        assert isProposerBefore, "Unauthorized proposal";
+        assert isDelaySufficient, "Minimum delay violation";
```

I believe that this is equivalent, but in my opinion it expresses the property more explicitly and clearly. What do you think?

```suggestion
    bool success = !lastRevert;

    assert success <=> stateBefore == UNSET() && isDelaySufficient && isProposerBefore, "Unexpected revert";

    if (success) {
        assert state(id) == PENDING(), "State transition violation";
```